Add CACHE_SIZE environment variable to the available options

jparenas opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This PR fixes #648 by adding a new environmental variable: CACHE_SIZE. It defaults to 10000 (following the default value in PiHole) and can be changed to any positive number, including 0. This will allow for people to change the default cache size and let the change survive restarts (and avoiding messing with the config file directly).

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
#648 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
I manually tested the change and it works, and I added some test code following some tests that were already present.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] My code follows the code style of this project.
- [x ] My change requires a change to the documentation.
- [x ] I have updated the documentation accordingly.

dschaper commented:
Have you tested a cache size of 0 with DNSSEC enabled?

jparenas commented:
I haven't tested a cache size of 0 with DNSSEC enabled, but going from the multiple threads online it seems that it would not break anything. 

Also, while I have you here, I saw that the tests are broken as they depend on python:buster and it recently was moved from Python 3.8 to Python 3.9, so they fail on line 20 of Dockerfile_build. Should I submit a new pull request to fix that or can I do it here?

dschaper commented:
>  it seems that it would not break anything.

Make no assumptions. Test it and verify. `dnsmasq` reacts very badly to 0 cache and DNSSEC. I don't remember offhand if `dnsmasq` now automatically caches or if it still just refuses to start.

I don't know the testing, this isn't my main area but @lightswitch05 or @diginc should be able to help out.

jparenas commented:
You are absolutely right, dnsmasq refuses to start if cache-size is changed. I made a change to avoid it changing if DNSSEC is enabled and added a test for it.

dschaper commented on Dockerfile_build:
```diff
@@ -17,7 +17,7 @@ COPY Pipfile* /root/
 
 RUN pipenv install --system \
-    && sed -i 's|/bin/sh|/bin/bash|g' /usr/local/lib/python3.8/site-packages/testinfra/backend/docker.py
+    && sed -i 's|/bin/sh|/bin/bash|g' /usr/local/lib/python3.9/site-packages/testinfra/backend/docker.py
```
This needs to be set back to 3.8, the `DOCKERFILE_build` in the dev branch is now pinned to 3.8

jparenas replied:
Changed back to 3.8

dschaper commented on README.md:
```diff
@@ -118,6 +118,7 @@ There are other environment variables if you want to customize various things in
 | `INTERFACE: <NIC>`<br/> *Advanced/Optional* | The default works fine with our basic example docker run commands.  If you're trying to use DHCP with `--net host` mode then you may have to customize this or DNSMASQ_LISTENING.
 | `DNSMASQ_LISTENING: <local\|all\|NIC>`<br/> *Advanced/Optional* | `local` listens on all local subnets, `all` permits listening on internet origin subnets in addition to local.
 | `WEB_PORT: <PORT>`<br/> *Advanced/Optional* | **This will break the 'webpage blocked' functionality of Pi-hole** however it may help advanced setups like those running synology or `--net=host` docker argument.  This guide explains how to restore webpage blocked functionality using a linux router DNAT rule: [Alternative Synology installation method](https://discourse.pi-hole.net/t/alternative-sy ...
+| `CACHE_SIZE: <size>`<br/> *Advanced/Optional: Default: '10000'* | Set the cache size for dnsmasq. Useful for increasing the default cache size or to set it to 0.
```
Need to have a note in here that you can not set it to 0 when using DNSSEC.

jparenas replied:
Added the note about DNSSEC. Thanks for bringing it up!

jparenas commented:
Fixed the requested changes, thanks for bringing them up!

dschaper commented on bash_functions.sh:
```diff
@@ -193,6 +194,32 @@ setup_dnsmasq_hostnames() {
+        return
+    fi
+
+    if ! echo $1 | grep -q '^[0-9][0-9]*$' ; then
```
Nitpick: Is there a reason for the regex like this or would `^[0-9]+$` work? Not marking as a change request, but if there's no objection to simplifying it then would be nice to have.

jparenas replied:
That's a nice find. I copied this line from https://github.com/pi-hole/docker-pi-hole/blob/f4814935b9c0c531c415b3c9ae58af97466f30cd/bash_functions.sh#L258, but looking closely at it it definitely would be better to simplify it. Changed.

jparenas commented:
All the requested changes have been made so please let me know if there are any other fixes that should be made to the PR. Thank you for your time and help!

PromoFaux commented:
@jparenas, apologies for the long time to get back to you. I merged in latest `dev` changes, but it seems that the tests you added are now failing, please take a look

jparenas commented:
Hi @PromoFaux, sorry for taking a while to get back to you. I updated the branch with a working version (apparently the upstream PiHole install script did some changes that shadowed the CACHE_SIZE env variable. I had to change the variable name because of that. That took a while to figure out) and rebased the branch to the latest commit.

Thank you for taking a look at it after so long!

PromoFaux reviewed (approved):
Re-approve after final sanity check before merge
